src/systems/filecoin_markets/storage_market/storage_market_actor.go

```diff
@@ -0,0 +1,110 @@
+package storage_market
+
+import addr "github.com/filecoin-project/specs/systems/filecoin_vm/actor/address"
+import deal "github.com/filecoin-project/specs/systems/filecoin_markets/deal"
+import actor "github.com/filecoin-project/specs/systems/filecoin_vm/actor"
+
+func (sma *StorageMarketActor_I) WithdrawBalance(balance actor.TokenAmount) {
+	var msgSender addr.Address // TODO replace this from VM runtime
+
+	if balance < 0 {
+		panic("Error")
+	}
+
+	senderBalance, found := sma.Balances()[msgSender]
+	if !found {
+		panic("Error") // TODO replace with fatal
+	}
+
+	if senderBalance.Available() < balance {
+		panic("Error") // TODO replace with fatal
+	}
+
+	senderBalance.Impl().Available_ = senderBalance.Available() - balance
+	sma.Balances()[msgSender] = senderBalance
+
+	// TODO send funds to msgSender with `send` in VM runtime
+}
+
+func (sma *StorageMarketActor_I) AddBalance(balance actor.TokenAmount) {
+	var msgSender addr.Address // TODO replace this
+
+	// TODO subtract balance from msgSender
+	// TODO add balance to StorageMarketActor
+	if balance < 0 {
+		panic("Error")
+	}
+
+	senderBalance, found := sma.Balances()[msgSender]
+	if found {
+		senderBalance.Impl().Available_ = senderBalance.Available() + balance
+		sma.Balances()[msgSender] = senderBalance
+	} else {
+		sma.Balances()[msgSender] = &StorageParticipantBalance_I{
+			Locked_:    0,
+			Available_: balance,
+		}
+	}
+}
+
+func (sma *StorageMarketActor_I) CheckLockedBalance(participantAddr addr.Address) actor.TokenAmount {
+	var msgSender addr.Address // TODO replace this
+
+	return sma.Balances()[msgSender].Locked()
+}
+
+func (sma *StorageMarketActor_I) PublishStorageDeal(newStorageDeals []deal.StorageDeal) []PublishStorageDealResponse {
+	l := len(newStorageDeals)
+	response := make([]PublishStorageDealResponse, l)
+	for i, newDeal := range newStorageDeals {
+		if sma.verifyStorageDeal(newDeal) {
+			id := sma.generateStorageDealID(newDeal)
+			sma.Deals()[id] = newDeal
+			response[i] = PublishStorageDealSuccess
+		} else {
+			response[i] = PublishStorageDealError
+		}
+	}
+
+	return response
+}
+
+func (sma *StorageMarketActor_I) verifyStorageDeal(d deal.StorageDeal) bool {
+	// TODO verify proposal or deal has not expired
+	// TODO verify client and provider signature
+	// TODO verfiy minimum StoragePrice and StorageCollateral
```

Comment:
we should also verify that the deal expiry is larger than some particular min amount of time and that the deal is of a particular size

Reply:
yes, minimum storage collateral/deal size are in the works